lib/internal/modules/esm/resolve.js

```diff
@@ -414,16 +451,22 @@ function resolvePackageTargetString(
   if (RegExpPrototypeExec(invalidSegmentRegEx, subpath) !== null) {
     const request = pattern ?
       StringPrototypeReplace(match, '*', () => subpath) : match + subpath;
-    throwInvalidSubpath(request, packageJSONUrl, internal, base);
+    if (RegExpPrototypeExec(deprecatedInvalidSegmentRegEx, subpath) === null) {
+      if (!isPathMap) {
+        const resolvedTarget = pattern ?
+          RegExpPrototypeSymbolReplace(patternRegEx, target, () => subpath) :
+          target;
+        emitInvalidSegmentDeprecation(resolvedTarget, request, match,
+                                      packageJSONUrl, base);
+      }
+    } else {
+      throwInvalidSubpath(request, match, packageJSONUrl, internal, base);
+    }
```

Comment:
This code has a lot of the same/similar code as the block above (both checking `deprecatedInvalidSegmentRegEx`, getting `resolvedTarget`, emitting the same deprecation, etc). Would it be worthwhile to abstract them into a common helper?

Reply:
Again, this isn't possible with the line limit.

Reply:
They are different codepaths and different error messages. Also this will simplify again once the deprecation is done, so I don't think it's worth combining.

Reply:
Ah, sorry it looked the same. Okay! Make sense that it's temp 🙂